---
title: "Resources"
---

# R programming
This workshop will utilize the R programmling language through the RStudio user interface. We will perform these activities on the RStudio Server, available through the [Posit Cloud](https://posit.cloud/) platform. While the *Introduction to R programming* (Days 1-2) is is designed for all skill levels, having a basic understanding of R programming may enhance your learning experience. If you are new to these areas, consider reviewing one or a few of the following introductory materials prior to the workshop. Over the course of the workshop, teaching assistants ([TAs](authors.qmd)) will be working with small groups of attendees during the hands-on exercises. Feel free to direct your questions to the them, the instructors, or the Slack channel.

-   *Posit Cloud account*: You should have received access to Posit Cloud through a link to your email address for this workshop. Logging into the workspace, using this email address, will provide access to all course materials and allow for data analysis through RStudio Server on your internet browser.
-   *Laptop computer*: Please bring your own laptop with internet capability. Since the workshop will take place through Posit Cloud, the interface will be accessed through an internet browser. Google Chrome is recommended.

## On your own (Outside of the course)
-   *R*: In order to utilize the RStudio Desktop application from your laptop (and not through the browser), you will first need to install the [latest version of R](https://www.r-project.org/).
-   *RStudio Desktop* can be downloaded [here](https://posit.co/download/rstudio-desktop/). Note that this functions in the same manner as the RStudio Server by Posit Cloud, but utilizes the storage and compute of the laptop machine itself. Laptops vary in memory and storage space, which can dictate the speed and capability of data analysis, and this may result in slower processes with large datasets.

# Sequencing analysis
-	*Integrated Genomics Viewer* (IGV): IGV will be used to visually explore DNA and RNA sequencing data. IGV is available as a [web application](https://igv.org/app/) or can be [downloaded to your computer](https://igv.org/doc/desktop/#DownloadPage/) (recommended 2.17.2 or greater) to visualize sequencing reads.

# Single cell RNA sequencing
-	*Loupe Browser*: The Loupe Browser will be used to facilitate exploration of single cell RNA sequencing (scRNAseq) data, and can be downloaded (after agreeing to the 10X License Agreement) [here](https://www.10xgenomics.com/support/software/loupe-browser/latest).

# Spatial transcriptomics
For sample pre-processing, the 10X Genomics [Visium HD Analysis with spaceranger count](https://www.10xgenomics.com/support/cn/software/space-ranger/3.0/analysis/running-pipelines/count-visium-hd) explains how to run SpaceRanger with automatic and manual alignment. For manual alignment of the Visium HD CytAssist image, please make sure to download [Loupe Browser 8.1.2](https://www.10xgenomics.com/support/software/loupe-browser/downloads).

-	To get started with how to perform analysis, visualization, and integration of Visium HD spatial datasets with Seurat, check this [vignette](https://satijalab.org/seurat/articles/visiumhd_analysis_vignette), which also includes information on how to run [Banksy](https://www.nature.com/articles/s41588-024-01664-3) unsupervised neighborhood clustering
-	For information on how to construct single-cell trajectories/pseudotime trajectories using Monocle3, check this [vignette](https://cole-trapnell-lab.github.io/monocle3/docs/trajectories/)
-	For using InferCNV with the Visium HD data, this [vignette](https://github.com/broadinstitute/inferCNV/wiki) illustrates inferring copy number alterations from tumor single cell RNA-Seq data
-	For nuclei Segmentation and Custom Binning of Visium HD Gene Expression Data, refer to this [vignette](https://www.10xgenomics.com/analysis-guides/segmentation-visium-hd) by 10X. This is the only step that requires working in Python, all the rest of the scripts work with Rstudio. To convert the Seurat object (R) to work with Python (AnnData) follow this [vignette](https://mojaveazure.github.io/seurat-disk/articles/convert-anndata.html).


### Other workshops and vignettes
-   [Single Cell Best Practices](https://www.sc-best-practices.org/preamble.html)
-   [RNA-seq Bioinformatics](https://rnabio.org/course/)
-   [Precision Medicine Bioinformatics](https://pmbio.org/)
-   [Orchestrating Single-Cell Analysis with Bioconductor](https://bioconductor.org/books/release/OSCA/)
-   [R for Data Science](https://r4ds.hadley.nz/)
