<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Reading, navigating, and plotting data – CRI Bioinformatics Workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../course/test_newfile.html" rel="next">
<link href="../course/3_clustering.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark-8ef56b68f8fa1e9d2ba328e99e439f80.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-083966fab483341b80272ffbd41d3d84.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark-50e486611be19e4f29225c18ed88561d.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../site_libs/bootstrap/bootstrap-083966fab483341b80272ffbd41d3d84.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link href="../site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="../site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="../site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="../site_libs/datatables-binding-0.33/datatables.js"></script><script src="../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><link href="../site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="../site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="../site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="../site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="../site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script><link href="../site_libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="../site_libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="../site_libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="../site_libs/selectize-0.12.0/selectize.min.js"></script><link rel="stylesheet" href="../styles.css">
</head>
<body class="nav-sidebar docked nav-fixed quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../img/cri_logo.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">CRI Bioinformatics Workshop</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../course.html"> 
<span class="menu-text">Course</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../authors.html"> 
<span class="menu-text">Faculty</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../resources.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://twitter.com/CancerResearch" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/Cancer-Research-Institute/CRI_bioinformatics_workshop">
            Source Code
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../course/1_introR.html">Intro to R</a></li><li class="breadcrumb-item"><a href="../course/4_qctrouble.html">Reading, navigating, and plotting data</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Intro to R</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course/1_introR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course/2_basicplotting.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic plotting and statistics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course/3_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering concepts and correlation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course/4_qctrouble.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Reading, navigating, and plotting data</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Bulk RNA Analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">RNA Bio website</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-01-inputs/0001/05/01/RNAseq_Data/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Bulk RNAseq dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-02-alignment/0002/05/02/AlignVis_IGV/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Intro to IGV</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-03-expression/0003/03/03/Differential_Expression-DESeq2/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Differential Expression Analysis (DESeq2)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course/test_newfile.html" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Code blocks for DE</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-03-expression/0003/04/02/DE_Visualization-DESeq2/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Differential Expression Visualization (DESeq2)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-03-expression/0003/04/03/DE_Visualization_AdvancedR/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Differential Expression Visualization (Advanced R)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-03-expression/0003/05/01/DE_Pathway_Analysis/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Pathway Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-03-expression/0003/06/02/Batch-Correction/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Batch Correction</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Single Cell RNA Sequencing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/01/02/scRNA_Data/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Introduction to scRNAseq</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/02/01/QA_clustering/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">QA/QC and Clustering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/03/01/Cell_annotation/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Cell Type Annotation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/04/01/DE_analysis/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Differential Expression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/05/01/Gene_set_enrichment/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Gene Set Enrichment</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/06/01/Cancer_cell_identification/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Cancer Cell Identification</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/07/01/Differentiation_trajectory/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">Trajectory Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="https://rnabio.org/module-08-scrna/0008/08/01/TCR_BCR_analysis/" class="sidebar-item-text sidebar-link" target="_blank">
 <span class="menu-text">TCR/BCR Repertoire Analysis</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Spatial Transcriptomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course/spatial/01_spatial_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Intro to Spatial Data</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#where-should-you-start-with-a-new-dataset" id="toc-where-should-you-start-with-a-new-dataset" class="nav-link active" data-scroll-target="#where-should-you-start-with-a-new-dataset">Where should you start with a new dataset?</a></li>
  <li>
<a href="#common-challenges" id="toc-common-challenges" class="nav-link" data-scroll-target="#common-challenges">Common challenges</a>
  <ul>
<li><a href="#nan-and-missing-data" id="toc-nan-and-missing-data" class="nav-link" data-scroll-target="#nan-and-missing-data"><em>NaN</em> and missing data</a></li>
  <li><a href="#data-organization-and-structure-factors" id="toc-data-organization-and-structure-factors" class="nav-link" data-scroll-target="#data-organization-and-structure-factors">Data organization and structure: Factors</a></li>
  <li><a href="#missing-packages-or-dependencies" id="toc-missing-packages-or-dependencies" class="nav-link" data-scroll-target="#missing-packages-or-dependencies">Missing packages or dependencies</a></li>
  </ul>
</li>
  <li>
<a href="#where-do-i-go-for-help" id="toc-where-do-i-go-for-help" class="nav-link" data-scroll-target="#where-do-i-go-for-help">Where do I go for help?</a>
  <ul>
<li><a href="#stack-exchange" id="toc-stack-exchange" class="nav-link" data-scroll-target="#stack-exchange">Stack Exchange</a></li>
  <li><a href="#chatgpt" id="toc-chatgpt" class="nav-link" data-scroll-target="#chatgpt">ChatGPT</a></li>
  <li><a href="#good-practices" id="toc-good-practices" class="nav-link" data-scroll-target="#good-practices">Good practices</a></li>
  <li>
<a href="#publicly-accessible-resources" id="toc-publicly-accessible-resources" class="nav-link" data-scroll-target="#publicly-accessible-resources">Publicly accessible resources</a>
  <ul class="collapse">
<li><a href="#github" id="toc-github" class="nav-link" data-scroll-target="#github">Github</a></li>
  </ul>
</li>
  </ul>
</li>
  <li><a href="#how-do-i-apply-the-code-ive-learned-to-my-own-data" id="toc-how-do-i-apply-the-code-ive-learned-to-my-own-data" class="nav-link" data-scroll-target="#how-do-i-apply-the-code-ive-learned-to-my-own-data">How do I apply the code I’ve learned to my own data?</a></li>
  <li>
<a href="#additional-practice" id="toc-additional-practice" class="nav-link" data-scroll-target="#additional-practice">Additional practice</a>
  <ul>
<li><a href="#recreate-a-paper-figure-from-scratch" id="toc-recreate-a-paper-figure-from-scratch" class="nav-link" data-scroll-target="#recreate-a-paper-figure-from-scratch">Recreate a Paper Figure from Scratch</a></li>
  <li>
<a href="#create-a-custom-heatmap" id="toc-create-a-custom-heatmap" class="nav-link" data-scroll-target="#create-a-custom-heatmap">Create a Custom Heatmap</a>
  <ul class="collapse">
<li><a href="#setup-toy-data" id="toc-setup-toy-data" class="nav-link" data-scroll-target="#setup-toy-data">Setup Toy Data</a></li>
  <li>
<a href="#approach-1---ggplot2" id="toc-approach-1---ggplot2" class="nav-link" data-scroll-target="#approach-1---ggplot2">Approach 1 - <code>ggplot2</code></a>
  <ul class="collapse">
<li><a href="#perform-clustering" id="toc-perform-clustering" class="nav-link" data-scroll-target="#perform-clustering">Perform clustering</a></li>
  <li><a href="#metadata-plot" id="toc-metadata-plot" class="nav-link" data-scroll-target="#metadata-plot">Metadata plot</a></li>
  <li><a href="#add-dendrograms" id="toc-add-dendrograms" class="nav-link" data-scroll-target="#add-dendrograms">Add Dendrograms</a></li>
  <li><a href="#put-it-all-together" id="toc-put-it-all-together" class="nav-link" data-scroll-target="#put-it-all-together">Put it all together</a></li>
  <li><a href="#bonus-challenge" id="toc-bonus-challenge" class="nav-link" data-scroll-target="#bonus-challenge">Bonus Challenge</a></li>
  </ul>
</li>
  <li><a href="#approach-2---live-recreation-using-ggplot2" id="toc-approach-2---live-recreation-using-ggplot2" class="nav-link" data-scroll-target="#approach-2---live-recreation-using-ggplot2">Approach 2 - Live Recreation using <code>ggplot2</code></a></li>
  <li><a href="#approach-3---complexheatmap" id="toc-approach-3---complexheatmap" class="nav-link" data-scroll-target="#approach-3---complexheatmap">Approach 3 - <code>ComplexHeatmap</code></a></li>
  </ul>
</li>
  </ul>
</li>
  </ul><div class="quarto-other-links"><h2>Other Links</h2><ul><li><a href="https://github.com/Cancer-Research-Institute/CRI_bioinformatics_workshop/blob/main/course/slides/CRI%20Bioinformatics%20Bootcamp%202025%20Day%202PM.pdf"><i class="bi bi-link-45deg"></i>Lecture Slides</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../course/1_introR.html">Intro to R</a></li><li class="breadcrumb-item"><a href="../course/4_qctrouble.html">Reading, navigating, and plotting data</a></li></ol></nav><div class="quarto-title">
<h1 class="title">Reading, navigating, and plotting data</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="where-should-you-start-with-a-new-dataset" class="level1"><h1>Where should you start with a new dataset?</h1>
<p>The key to accomplishing any analysis is to start by understand what your data looks like and how it’s organized. This might include:</p>
<ul>
<li>What type of files are you working with?</li>
<li>How do they get loaded into R?</li>
<li>What is the size of the dataset?</li>
<li>What types of questions can I ask of the data?</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">variant_file</span> <span class="op">&lt;-</span> <span class="st">"/cloud/project/data/single_cell_rna/cancer_cell_id/mcb6c-exome-somatic.variants.annotated.clean.filtered.tsv"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">variant_file</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/cloud/project/data/single_cell_rna/cancer_cell_id/mcb6c-exome-somatic.variants.annotated.clean.filtered.tsv"</code></pre>
</div>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># File is a "tsv" file -&gt; Tab-delimited file</span></span>
<span><span class="va">read_tsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">variant_file</span>, sep <span class="op">=</span> <span class="st">'\t'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now we can explore the file we just loaded:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Look at the file: View(), head(), or click on it to the right</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  CHROM     POS REF ALT                    set NORMAL.GT NORMAL.AD NORMAL.AF
1  chr1 4347274   T   A mutect-varscan-strelka       T/T     122,0         0
2  chr1 4347386   C   A mutect-varscan-strelka       C/C      96,0         0
3  chr1 4830263   C   T mutect-varscan-strelka       C/C     195,0         0
4  chr1 5084567   C   T mutect-varscan-strelka       C/C      61,0         0
5  chr1 6247967   T   G mutect-varscan-strelka       T/T      79,0         0
6  chr1 6249819   T   A mutect-varscan-strelka       T/T     189,0         0
  NORMAL.DP TUMOR.GT TUMOR.AD TUMOR.AF TUMOR.DP
1       122      T/A    44,48  0.52174       92
2        96      C/A    42,42  0.50000       84
3       195      C/T    79,64  0.44755      143
4        62      C/T    26,19  0.42222       45
5        79      T/G    24,18  0.42857       42
6       189      T/A    58,57  0.49565      115
                           Consequence  SYMBOL Feature_type
1                     missense_variant     Rp1   Transcript
2                     missense_variant     Rp1   Transcript
3 splice_region_variant&amp;intron_variant  Lypla1   Transcript
4 splice_region_variant&amp;intron_variant Atp6v1h   Transcript
5                       intron_variant  Rb1cc1   Transcript
6                     missense_variant  Rb1cc1   Transcript
                Feature cDNA_position CDS_position Protein_position Amino_acids
1  ENSMUST00000027032.5          3741         3614             1205         N/I
2  ENSMUST00000027032.5          3629         3502             1168         V/F
3 ENSMUST00000027036.10                                                        
4 ENSMUST00000044369.12                                                        
5 ENSMUST00000027040.12                                                        
6 ENSMUST00000027040.12          3928         3461             1154         V/D
   Codons NORMAL.REF.AD NORMAL.ALT.AD TUMOR.REF.AD TUMOR.ALT.AD
1 aAt/aTt           122             0           44           48
2 Gtt/Ttt            96             0           42           42
3                   195             0           79           64
4                    61             0           26           19
5                    79             0           24           18
6 gTt/gAt           189             0           58           57</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Understand the variables and data structure: typeof(), str(), colnames()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/typeof.html">typeof</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "list"</code></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   10427 obs. of  26 variables:
 $ CHROM           : chr  "chr1" "chr1" "chr1" "chr1" ...
 $ POS             : int  4347274 4347386 4830263 5084567 6247967 6249819 6264623 6842242 6842967 6844056 ...
 $ REF             : chr  "T" "C" "C" "C" ...
 $ ALT             : chr  "A" "A" "T" "T" ...
 $ set             : chr  "mutect-varscan-strelka" "mutect-varscan-strelka" "mutect-varscan-strelka" "mutect-varscan-strelka" ...
 $ NORMAL.GT       : chr  "T/T" "C/C" "C/C" "C/C" ...
 $ NORMAL.AD       : chr  "122,0" "96,0" "195,0" "61,0" ...
 $ NORMAL.AF       : num  0 0 0 0 0 0 0 0 0 0 ...
 $ NORMAL.DP       : int  122 96 195 62 79 189 83 60 86 77 ...
 $ TUMOR.GT        : chr  "T/A" "C/A" "C/T" "C/T" ...
 $ TUMOR.AD        : chr  "44,48" "42,42" "79,64" "26,19" ...
 $ TUMOR.AF        : num  0.522 0.5 0.448 0.422 0.429 ...
 $ TUMOR.DP        : int  92 84 143 45 42 115 63 33 55 43 ...
 $ Consequence     : chr  "missense_variant" "missense_variant" "splice_region_variant&amp;intron_variant" "splice_region_variant&amp;intron_variant" ...
 $ SYMBOL          : chr  "Rp1" "Rp1" "Lypla1" "Atp6v1h" ...
 $ Feature_type    : chr  "Transcript" "Transcript" "Transcript" "Transcript" ...
 $ Feature         : chr  "ENSMUST00000027032.5" "ENSMUST00000027032.5" "ENSMUST00000027036.10" "ENSMUST00000044369.12" ...
 $ cDNA_position   : chr  "3741" "3629" "" "" ...
 $ CDS_position    : chr  "3614" "3502" "" "" ...
 $ Protein_position: chr  "1205" "1168" "" "" ...
 $ Amino_acids     : chr  "N/I" "V/F" "" "" ...
 $ Codons          : chr  "aAt/aTt" "Gtt/Ttt" "" "" ...
 $ NORMAL.REF.AD   : int  122 96 195 61 79 189 83 60 86 77 ...
 $ NORMAL.ALT.AD   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ TUMOR.REF.AD    : int  44 42 79 26 24 58 35 11 24 21 ...
 $ TUMOR.ALT.AD    : int  48 42 64 19 18 57 28 22 31 22 ...</code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "CHROM"            "POS"              "REF"              "ALT"             
 [5] "set"              "NORMAL.GT"        "NORMAL.AD"        "NORMAL.AF"       
 [9] "NORMAL.DP"        "TUMOR.GT"         "TUMOR.AD"         "TUMOR.AF"        
[13] "TUMOR.DP"         "Consequence"      "SYMBOL"           "Feature_type"    
[17] "Feature"          "cDNA_position"    "CDS_position"     "Protein_position"
[21] "Amino_acids"      "Codons"           "NORMAL.REF.AD"    "NORMAL.ALT.AD"   
[25] "TUMOR.REF.AD"     "TUMOR.ALT.AD"    </code></pre>
</div>
</div>
</section><section id="common-challenges" class="level1"><h1>Common challenges</h1>
<section id="nan-and-missing-data" class="level2"><h2 class="anchored" data-anchor-id="nan-and-missing-data">
<em>NaN</em> and missing data</h2>
<p>Dealing with missing data and NaN (Not a Number) values is a common challenge in R programming. These values can affect the results of your analyses and visualizations. It’s essential to handle missing data appropriately, either by imputing them or excluding them from the analysis, depending on the context.</p>
<p><strong>Example</strong>:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Creating a dataframe with missing values</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="cn">NA</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>  y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="cn">NA</span>, <span class="fl">7</span>, <span class="fl">8</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check for missing values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Understanding missing or sparse information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    CHROM                POS                REF                ALT           
 Length:10427       Min.   :  2107407   Length:10427       Length:10427      
 Class :character   1st Qu.: 38008521   Class :character   Class :character  
 Mode  :character   Median : 73553346   Mode  :character   Mode  :character  
                    Mean   : 75250958                                        
                    3rd Qu.:107567817                                        
                    Max.   :195169782                                        
                    NA's   :1                                                
     set             NORMAL.GT          NORMAL.AD           NORMAL.AF        
 Length:10427       Length:10427       Length:10427       Min.   :0.0000000  
 Class :character   Class :character   Class :character   1st Qu.:0.0000000  
 Mode  :character   Mode  :character   Mode  :character   Median :0.0000000  
                                                          Mean   :0.0001637  
                                                          3rd Qu.:0.0000000  
                                                          Max.   :0.0408200  
                                                          NA's   :1          
   NORMAL.DP        TUMOR.GT           TUMOR.AD            TUMOR.AF      
 Min.   :  31.0   Length:10427       Length:10427       Min.   :0.05063  
 1st Qu.:  67.0   Class :character   Class :character   1st Qu.:0.41860  
 Median :  97.0   Mode  :character   Mode  :character   Median :0.48247  
 Mean   : 118.3                                         Mean   :0.48403  
 3rd Qu.: 148.0                                         3rd Qu.:0.54217  
 Max.   :1351.0                                         Max.   :1.00000  
 NA's   :2                                              NA's   :1        
    TUMOR.DP      Consequence           SYMBOL          Feature_type      
 Min.   : 31.00   Length:10427       Length:10427       Length:10427      
 1st Qu.: 45.00   Class :character   Class :character   Class :character  
 Median : 65.00   Mode  :character   Mode  :character   Mode  :character  
 Mean   : 79.68                                                           
 3rd Qu.: 98.00                                                           
 Max.   :813.00                                                           
 NA's   :2                                                                
   Feature          cDNA_position      CDS_position       Protein_position  
 Length:10427       Length:10427       Length:10427       Length:10427      
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            
                                                                            
                                                                            
                                                                            
 Amino_acids           Codons          NORMAL.REF.AD    NORMAL.ALT.AD    
 Length:10427       Length:10427       Min.   :  30.0   Min.   :0.00000  
 Class :character   Class :character   1st Qu.:  67.0   1st Qu.:0.00000  
 Mode  :character   Mode  :character   Median :  97.0   Median :0.00000  
                                       Mean   : 118.2   Mean   :0.01928  
                                       3rd Qu.: 148.0   3rd Qu.:0.00000  
                                       Max.   :1350.0   Max.   :2.00000  
                                       NA's   :2        NA's   :2        
  TUMOR.REF.AD     TUMOR.ALT.AD   
 Min.   :  0.00   Min.   :  4.00  
 1st Qu.: 23.00   1st Qu.: 21.00  
 Median : 34.00   Median : 31.00  
 Mean   : 41.68   Mean   : 37.96  
 3rd Qu.: 52.00   3rd Qu.: 47.00  
 Max.   :691.00   Max.   :229.00  
 NA's   :2        NA's   :2       </code></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">read_tsv</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html">complete.cases</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span>,<span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      CHROM      POS   REF  ALT            set   NORMAL.GT NORMAL.AD NORMAL.AF
9862   chr9 53503887 AACAC  AAC mutect-varscan AACAC/AACAC      &lt;NA&gt;     0.014
10272  &lt;NA&gt;       NA  &lt;NA&gt; &lt;NA&gt;           &lt;NA&gt;        &lt;NA&gt;      &lt;NA&gt;        NA
      NORMAL.DP TUMOR.GT TUMOR.AD TUMOR.AF TUMOR.DP    Consequence SYMBOL
9862         NA  AACAC/.     &lt;NA&gt;    0.184       NA intron_variant    Atm
10272        NA     &lt;NA&gt;     &lt;NA&gt;       NA       NA           &lt;NA&gt;   &lt;NA&gt;
      Feature_type              Feature cDNA_position CDS_position
9862    Transcript ENSMUST00000232179.1                           
10272         &lt;NA&gt;                 &lt;NA&gt;          &lt;NA&gt;         &lt;NA&gt;
      Protein_position Amino_acids Codons NORMAL.REF.AD NORMAL.ALT.AD
9862                                                 NA            NA
10272             &lt;NA&gt;        &lt;NA&gt;   &lt;NA&gt;            NA            NA
      TUMOR.REF.AD TUMOR.ALT.AD
9862            NA           NA
10272           NA           NA</code></pre>
</div>
</div>
</section><section id="data-organization-and-structure-factors" class="level2"><h2 class="anchored" data-anchor-id="data-organization-and-structure-factors">Data organization and structure: Factors</h2>
<p>Columns that contain strings are automatically read in as character vectors and are arranged alphanumericall. <em>Factors</em> assign a logical order to a series of samples.</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co"># Organize the chromosomes to the correct order</span></span>
<span><span class="co"># If CHROM is a character vector, the chromosomes are not ordered properly</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span> <span class="co"># Shows that CHROM variable is a character vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   10427 obs. of  26 variables:
 $ CHROM           : chr  "chr1" "chr1" "chr1" "chr1" ...
 $ POS             : int  4347274 4347386 4830263 5084567 6247967 6249819 6264623 6842242 6842967 6844056 ...
 $ REF             : chr  "T" "C" "C" "C" ...
 $ ALT             : chr  "A" "A" "T" "T" ...
 $ set             : chr  "mutect-varscan-strelka" "mutect-varscan-strelka" "mutect-varscan-strelka" "mutect-varscan-strelka" ...
 $ NORMAL.GT       : chr  "T/T" "C/C" "C/C" "C/C" ...
 $ NORMAL.AD       : chr  "122,0" "96,0" "195,0" "61,0" ...
 $ NORMAL.AF       : num  0 0 0 0 0 0 0 0 0 0 ...
 $ NORMAL.DP       : int  122 96 195 62 79 189 83 60 86 77 ...
 $ TUMOR.GT        : chr  "T/A" "C/A" "C/T" "C/T" ...
 $ TUMOR.AD        : chr  "44,48" "42,42" "79,64" "26,19" ...
 $ TUMOR.AF        : num  0.522 0.5 0.448 0.422 0.429 ...
 $ TUMOR.DP        : int  92 84 143 45 42 115 63 33 55 43 ...
 $ Consequence     : chr  "missense_variant" "missense_variant" "splice_region_variant&amp;intron_variant" "splice_region_variant&amp;intron_variant" ...
 $ SYMBOL          : chr  "Rp1" "Rp1" "Lypla1" "Atp6v1h" ...
 $ Feature_type    : chr  "Transcript" "Transcript" "Transcript" "Transcript" ...
 $ Feature         : chr  "ENSMUST00000027032.5" "ENSMUST00000027032.5" "ENSMUST00000027036.10" "ENSMUST00000044369.12" ...
 $ cDNA_position   : chr  "3741" "3629" "" "" ...
 $ CDS_position    : chr  "3614" "3502" "" "" ...
 $ Protein_position: chr  "1205" "1168" "" "" ...
 $ Amino_acids     : chr  "N/I" "V/F" "" "" ...
 $ Codons          : chr  "aAt/aTt" "Gtt/Ttt" "" "" ...
 $ NORMAL.REF.AD   : int  122 96 195 61 79 189 83 60 86 77 ...
 $ NORMAL.ALT.AD   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ TUMOR.REF.AD    : int  44 42 79 26 24 58 35 11 24 21 ...
 $ TUMOR.ALT.AD    : int  48 42 64 19 18 57 28 22 31 22 ...</code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">read_tsv</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CHROM</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># make CHROM a factor</span></span>
<span><span class="va">read_tsv</span><span class="op">$</span><span class="va">CHROM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">$</span><span class="va">CHROM</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">'X'</span>,<span class="st">'Y'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">read_tsv</span><span class="op">)</span> <span class="co"># shows that CHROM variable is a factor with a set order of character strings</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   10427 obs. of  26 variables:
 $ CHROM           : Factor w/ 24 levels "chr1","chr2",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ POS             : int  4347274 4347386 4830263 5084567 6247967 6249819 6264623 6842242 6842967 6844056 ...
 $ REF             : chr  "T" "C" "C" "C" ...
 $ ALT             : chr  "A" "A" "T" "T" ...
 $ set             : chr  "mutect-varscan-strelka" "mutect-varscan-strelka" "mutect-varscan-strelka" "mutect-varscan-strelka" ...
 $ NORMAL.GT       : chr  "T/T" "C/C" "C/C" "C/C" ...
 $ NORMAL.AD       : chr  "122,0" "96,0" "195,0" "61,0" ...
 $ NORMAL.AF       : num  0 0 0 0 0 0 0 0 0 0 ...
 $ NORMAL.DP       : int  122 96 195 62 79 189 83 60 86 77 ...
 $ TUMOR.GT        : chr  "T/A" "C/A" "C/T" "C/T" ...
 $ TUMOR.AD        : chr  "44,48" "42,42" "79,64" "26,19" ...
 $ TUMOR.AF        : num  0.522 0.5 0.448 0.422 0.429 ...
 $ TUMOR.DP        : int  92 84 143 45 42 115 63 33 55 43 ...
 $ Consequence     : chr  "missense_variant" "missense_variant" "splice_region_variant&amp;intron_variant" "splice_region_variant&amp;intron_variant" ...
 $ SYMBOL          : chr  "Rp1" "Rp1" "Lypla1" "Atp6v1h" ...
 $ Feature_type    : chr  "Transcript" "Transcript" "Transcript" "Transcript" ...
 $ Feature         : chr  "ENSMUST00000027032.5" "ENSMUST00000027032.5" "ENSMUST00000027036.10" "ENSMUST00000044369.12" ...
 $ cDNA_position   : chr  "3741" "3629" "" "" ...
 $ CDS_position    : chr  "3614" "3502" "" "" ...
 $ Protein_position: chr  "1205" "1168" "" "" ...
 $ Amino_acids     : chr  "N/I" "V/F" "" "" ...
 $ Codons          : chr  "aAt/aTt" "Gtt/Ttt" "" "" ...
 $ NORMAL.REF.AD   : int  122 96 195 61 79 189 83 60 86 77 ...
 $ NORMAL.ALT.AD   : int  0 0 0 0 0 0 0 0 0 0 ...
 $ TUMOR.REF.AD    : int  44 42 79 26 24 58 35 11 24 21 ...
 $ TUMOR.ALT.AD    : int  48 42 64 19 18 57 28 22 31 22 ...</code></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">read_tsv</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">CHROM</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="missing-packages-or-dependencies" class="level2"><h2 class="anchored" data-anchor-id="missing-packages-or-dependencies">Missing packages or dependencies</h2>
<p>When running R code, you may encounter errors related to missing packages or dependencies. This occurs when you try to use functions or libraries that are not installed on your system. This will commonly look like a red error stating “There is no package…” Installing the required packages using <em>install.packages(“package_name”)</em> can resolve this issue.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Trying to use a function from an uninstalled package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Error: there is no package called 'ggplot2'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section></section><section id="where-do-i-go-for-help" class="level1"><h1>Where do I go for help?</h1>
<section id="stack-exchange" class="level2"><h2 class="anchored" data-anchor-id="stack-exchange">Stack Exchange</h2>
<p>Online communities like <a href="https://stackexchange.com/tour">Stack Exchange</a>, particularly the <a href="https://stackoverflow.com/questions">Stack Overflow</a> platform, are excellent resources for getting help with R programming. You can search for solutions to specific problems or ask questions if you’re facing challenges. If you attempt to Google what you’re trying to accomplish with your dataset, StackExchange often includes responses from others who may have gone through this effort before. With the open crowdsourcing of troubleshooting, these responses that work for others are “upvoted” so that you can try to adapt the code to work for your dataset. Importantly, take care to change the names of variables and file paths in any code that you try to implement from others.</p>
</section><section id="chatgpt" class="level2"><h2 class="anchored" data-anchor-id="chatgpt">ChatGPT</h2>
<p>ChatGPT is an AI assistant that can provide guidance and answer questions related to R programming. You can ask for clarification on concepts, debugging assistance, or advice on best practices. If you copy/paste an error into ChatGPT, it often tries to debug without any other preface. However, if you try to explain exactly what your code is trying to accomplish, it can be a helpful way to debug your help.</p>
</section><section id="good-practices" class="level2"><h2 class="anchored" data-anchor-id="good-practices">Good practices</h2>
<p>##Commenting your code Adding comments to your code is crucial for making it more understandable to yourself and others. Comments provide context and explanations for the code’s functionality, making it easier to troubleshoot and maintain.</p>
</section><section id="publicly-accessible-resources" class="level2"><h2 class="anchored" data-anchor-id="publicly-accessible-resources">Publicly accessible resources</h2>
<section id="github" class="level3"><h3 class="anchored" data-anchor-id="github">Github</h3>
<p>Github hosts numerous repositories containing R scripts, packages, and projects. Browsing through repositories and contributing to open-source projects can help you learn from others’ code and collaborate with the R community. This can be a direct way to make your code available upon publication, according to journal practices.</p>
</section></section></section><section id="how-do-i-apply-the-code-ive-learned-to-my-own-data" class="level1"><h1>How do I apply the code I’ve learned to my own data?</h1>
<p>Once you’ve learned R programming basics, applying the code to your own data involves understanding your data structure, identifying relevant functions and packages, and adapting example code to suit your specific analysis goals.</p>
</section><section id="additional-practice" class="level1"><h1>Additional practice</h1>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Read in the file "/cloud/project/data/bulk_rna/GSE48035_ILMN.Counts.SampleSubset.ProteinCodingGenes.tsv"</span></span>
<span><span class="co"># Describe the size, shape, and organization of this data file.</span></span>
<span></span>
<span></span>
<span><span class="co"># Example: Applying code to your own data</span></span>
<span><span class="co"># Load your dataset</span></span>
<span><span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"my_data.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Explore the structure of your data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">my_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Identify relevant variables and features</span></span>
<span><span class="co"># HINT: Use colnames() and tidyverse functions to clean up your data</span></span>
<span></span>
<span><span class="co"># Perform analysis or visualization using appropriate functions and packages</span></span>
<span><span class="co"># HINT: Use ggpubr functions to parallelize statistics across groups</span></span>
<span></span>
<span><span class="co"># Adapt example code to suit your data and analysis goals</span></span>
<span><span class="co"># HINT: Try different methods of plotting. Consider what important features you're trying to understand and show with this data.</span></span>
<span></span>
<span><span class="co"># Comment your code for clarity and future reference</span></span>
<span><span class="co"># HINT: Make your life easier when it comes time for reproducing your figures and filling in that "Data and Code Availability" section of your manuscript.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="recreate-a-paper-figure-from-scratch" class="level2"><h2 class="anchored" data-anchor-id="recreate-a-paper-figure-from-scratch">Recreate a Paper Figure from Scratch</h2>
<p>Here we will recreate a figure from <a href="https://www.science.org/doi/10.1126/science.abj4008">Schmidt et al (Science 2022)</a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/clipboard-2660320388.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:4in"></p>
</figure>
</div>
<p>First we will load the data from the supplemental table (<a href="https://github.com/crigrants/CRI_bioinformatics_workshop/raw/c66ec1f1da51a4868b76433a32d0915af3629a8a/course/data/science.abj4008_table_s2.xlsx">download here</a>):</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/">ggrepel</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">log2_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span><span class="st">"data/science.abj4008_table_s2.xlsx"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Always good practice to check the data to ensure it was loaded correctly:</span></span>
<span><span class="co"># head(log2_df)</span></span>
<span></span>
<span><span class="co"># The plot only contains the IL2 results from CRISPRa screens,</span></span>
<span><span class="co"># so we will filter down</span></span>
<span><span class="va">log2_df_filter</span> <span class="op">&lt;-</span> <span class="va">log2_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Cytokine</span> <span class="op">==</span> <span class="st">"IL2"</span>, <span class="va">CRISPRa_or_i</span> <span class="op">==</span> <span class="st">"CRISPRa"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select genes that we want to label based on top LFC</span></span>
<span><span class="va">ngenes_label</span> <span class="op">&lt;-</span> <span class="fl">16</span></span>
<span><span class="va">label_genes</span> <span class="op">&lt;-</span> <span class="va">log2_df_filter</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Screen_Version</span> <span class="op">==</span> <span class="st">"Primary"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># For ease only use primary donor</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">LFC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># Group by the sign (pos or neg) lfc</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">LFC</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># Sort by descending absolute lfc values</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">ngenes_label</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># Take the top n genes</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span> <span class="co"># Pull these out from the data frame</span></span>
<span></span>
<span><span class="co"># We need to pivot the Screen_Version variable (which is equivalent to Donor1 and Donor 2) into separate columns</span></span>
<span><span class="va">log2_df_filter_wide</span> <span class="op">&lt;-</span> <span class="va">log2_df_filter</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="st">"Hit_Type"</span><span class="op">)</span>, names_from <span class="op">=</span> <span class="st">"Screen_Version"</span>, values_from <span class="op">=</span> <span class="st">"LFC"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">Primary</span><span class="op">)</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">CD4_Supplement</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Reorder the factor so that we can plot the hits on top</span></span>
<span><span class="va">log2_df_filter_wide</span><span class="op">$</span><span class="va">Hit_Type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">log2_df_filter_wide</span><span class="op">$</span><span class="va">Hit_Type</span>,</span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Positive Hit"</span>, <span class="st">"Negative Hit"</span>, <span class="st">"NA"</span><span class="op">)</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Positive Hit"</span>, <span class="st">"Negative Hit"</span>, <span class="st">"Not a Hit"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Now we can plot:</span></span>
<span><span class="va">final_plot</span> <span class="op">&lt;-</span> <span class="va">log2_df_filter_wide</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># This will order by the factor so the hits are plotted on top</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">Hit_Type</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Primary</span>, y <span class="op">=</span> <span class="va">CD4_Supplement</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Hit_Type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html">stat_density_2d</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># For this geom text, we filter the original data to only include</span></span>
<span>  <span class="co"># genes that we want labeled from the above filtering</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">log2_df_filter_wide</span>, <span class="va">Gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">label_genes</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Gene</span>, colour <span class="op">=</span> <span class="va">Hit_Type</span><span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.36</span> <span class="op">*</span> <span class="fl">7</span>, show.legend <span class="op">=</span> <span class="cn">F</span>, max.overlaps <span class="op">=</span> <span class="cn">Inf</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Positive Hit"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"Not a Hit"</span> <span class="op">=</span> <span class="st">"grey80"</span>, <span class="st">"Negative Hit"</span> <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    <span class="co"># Place legend inside the plot</span></span>
<span>    legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.85</span>, <span class="fl">0.15</span><span class="op">)</span>,</span>
<span>    <span class="co"># Change vertical spacing between legend items</span></span>
<span>    legend.key.height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    <span class="co"># Place a box around the legend</span></span>
<span>    legend.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Donor 1, log2FoldChange"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Donor 2, log2FoldChange"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"IL-2 CRISPRa Screen"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Screen Hit"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co"># We can save our plot if desired</span></span>
<span><span class="co"># ggsave(final_plot, file = "my_pretty_plot.pdf", width = 5, height = 4)</span></span>
<span></span>
<span><span class="co"># And now view it</span></span>
<span><span class="va">final_plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
</section><section id="create-a-custom-heatmap" class="level2"><h2 class="anchored" data-anchor-id="create-a-custom-heatmap">Create a Custom Heatmap</h2>
<p>Here we will use the excellent R package <a href="https://patchwork.data-imaginist.com/"><code>patchwork</code></a> to compose a complex heatmap, that is both clustered and plotted next to relevant sample metadata and dendrograms.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="ggplot_heatmap.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:4in"></p>
</figure>
</div>
<section id="setup-toy-data" class="level3"><h3 class="anchored" data-anchor-id="setup-toy-data">Setup Toy Data</h3>
<p>NOTE: This is not important for understanding how to plot, but kept here in case you want to adjust some of these parameters to see what happens!</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load necessary libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate gene expression data</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">expression_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">50</span> <span class="op">*</span> <span class="fl">20</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">50</span>, dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Introduce differential expression for case vs control</span></span>
<span><span class="va">idx_case_genes</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span></span>
<span><span class="va">idx_conrol_genes</span> <span class="op">&lt;-</span> <span class="fl">21</span><span class="op">:</span><span class="fl">40</span></span>
<span><span class="va">idx_case_samples</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="va">idx_control_samples</span> <span class="op">&lt;-</span> <span class="fl">11</span><span class="op">:</span><span class="fl">20</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="va">idx_case_genes</span>, <span class="va">idx_case_samples</span><span class="op">]</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">expression_data</span><span class="op">[</span><span class="va">idx_case_genes</span>, <span class="va">idx_case_samples</span><span class="op">]</span> <span class="op">+</span> <span class="fl">3</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="va">idx_conrol_genes</span>, <span class="va">idx_control_samples</span><span class="op">]</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">expression_data</span><span class="op">[</span><span class="va">idx_conrol_genes</span>, <span class="va">idx_control_samples</span><span class="op">]</span> <span class="op">-</span> <span class="fl">3</span></span>
<span><span class="co"># Some drug specific effects</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">expression_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="op">-</span> <span class="fl">3</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">expression_data</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="op">-</span> <span class="fl">3</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="fl">21</span><span class="op">:</span><span class="fl">25</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">expression_data</span><span class="op">[</span><span class="fl">21</span><span class="op">:</span><span class="fl">25</span>, <span class="fl">11</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span> <span class="op">+</span> <span class="fl">3</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="fl">26</span><span class="op">:</span><span class="fl">30</span>, <span class="fl">16</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">expression_data</span><span class="op">[</span><span class="fl">26</span><span class="op">:</span><span class="fl">30</span>, <span class="fl">16</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span> <span class="op">+</span> <span class="fl">3</span></span>
<span></span>
<span><span class="co"># Create some metadata</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  Sample <span class="op">=</span> <span class="va">samples</span>,</span>
<span>  Condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Case"</span>, <span class="st">"Control"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>  Gender <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>, times <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>  Drug <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"DrugA"</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"DrugB"</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"DrugC"</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"DrugD"</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="approach-1---ggplot2" class="level3"><h3 class="anchored" data-anchor-id="approach-1---ggplot2">Approach 1 - <code>ggplot2</code>
</h3>
<p>We have a matrix of gene expression values:</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-b1cd96b95228713638ca" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-b1cd96b95228713638ca">{"x":{"filter":"none","vertical":false,"caption":"<caption>Expression Matrix<\/caption>","data":[["Gene1","Gene2","Gene3","Gene4","Gene5","Gene6","Gene7","Gene8","Gene9","Gene10","Gene11","Gene12","Gene13","Gene14","Gene15","Gene16","Gene17","Gene18","Gene19","Gene20","Gene21","Gene22","Gene23","Gene24","Gene25","Gene26","Gene27","Gene28","Gene29","Gene30","Gene31","Gene32","Gene33","Gene34","Gene35","Gene36","Gene37","Gene38","Gene39","Gene40","Gene41","Gene42","Gene43","Gene44","Gene45","Gene46","Gene47","Gene48","Gene49","Gene50"],[-0.96,-0.29,0.26,-1.15,0.2,3.03,3.09,4.12,1.78,4.27,2.26,1.87,2.28,3.25,3.15,2.69,2.05,2.35,4.22,3.2,-0.58,-0.9399999999999999,-0.2,-1.67,-0.48,-0.74,1.16,1.01,-0.07000000000000001,-1.14,0.9,0.85,0.73,0.74,-0.35,0.71,1.3,0.04,-0.98,0.79,0.79,-0.31,1.7,-0.79,0.35,-2.27,-0.16,1.13,-0.46,-0.9],[0.73,-0.8100000000000001,0.27,-1.74,-1.41,2.55,1.96,4.36,3.92,2.21,3.57,3.92,3.26,3.35,4.17,2.52,2.58,3.96,1.71,3.19,-0.03,0.47,1.02,0.27,0.23,0.75,1.22,0.38,-0.99,-0.16,1.74,-0.35,0.6899999999999999,1.22,0.79,-0.01,0.22,-0.89,0.44,-0.89,-0.85,-0.99,-0.65,1.05,-0.39,-0.07000000000000001,-0.46,0.54,0.93,-0.21],[0.62,-0.41,1.05,0.6,1.02,3.61,3.21,1.1,2.32,3.48,2.54,2.72,2.59,4.62,2.28,2.55,3.01,3.22,3.19,2.95,-1.5,0.37,0.52,-0.48,0.67,-0.76,0.39,-0.66,-1.72,1.16,0.6899999999999999,0.14,1.49,-1.63,0.13,-2.4,1.44,-0.88,-1.31,-0.88,-1.16,-1.98,-0.99,-0.15,0.91,0.41,-1.24,-0.64,1.93,0.41],[-1.29,2.64,0.49,0.85,1.09,3.23,3.07,2.02,1.69,5.46,2.33,3.91,3.96,4.61,4.84,3.7,4.22,1.88,3.67,4.22,0.23,-0.42,0.24,-0.55,-0.5,1.16,2.16,-1.71,-1.6,-1.04,0.32,-0.89,0.39,0.24,-0.43,-0.55,-1.32,0.68,2.16,-0.42,-1.36,-0.67,0.65,0.77,2.68,-1.37,0.06,-0.2,-1.26,-0.66],[-1.33,0.28,1.09,-1.64,-0.46,4.46,1.33,4.56,1.51,1.82,2.64,2.08,3.85,2.51,3.73,2.38,3.34,4.81,4.45,3.41,0.2,1.68,-0.96,0.66,-0.55,2.44,-0.3,-0.12,0.97,0.59,-1.21,-0.3,0.4,-1.69,-0.71,-0.25,-1,-0.92,0.44,-0.33,-0.27,1.28,0.3,1.27,1.03,0.67,-0.31,0.46,0.48,2.54],[2.46,4.37,3.14,2.22,4.88,0.55,-1.84,-0.55,-0.03,-2.16,4.77,3.76,2.52,2.53,3.33,2.03,2.47,3.06,3.1,4.17,0.39,-0.13,-0.38,1.28,-0.7,0.5600000000000001,0.3,0.22,0.92,-0.89,-0,-0.2,-2.23,0.48,1.62,-1.2,1.66,0.2,1.39,-0.86,0.6899999999999999,-0.58,-0.68,0.38,1.31,0.31,-0.5,1.25,-0.3,0.8100000000000001],[1.74,3.76,2.2,2.62,1.63,3.52,-1.21,0.19,0.09,-0.88,3.26,3.66,3.77,0.76,2.1,2.83,2.93,2.49,3.52,3.04,-0.91,0.91,-0.34,0.09,2.13,0.2,-0.97,-0.28,0.16,-1.11,-1.93,-0.47,0.53,-0.92,0.68,1.52,0.19,1.7,0.28,-0.57,-0.16,2.49,1.07,0.34,0.22,0.07000000000000001,0.31,1.51,0.62,0.09],[0.6,3.05,3.34,3.43,3.21,0.11,-1.99,1.51,0.35,0.43,3.77,3.41,1.02,1.91,4.06,5.12,2.49,3.41,5.04,2.81,1.37,-0.52,-0.72,-0.18,-0.4,0.83,-0.79,-0.9399999999999999,0.42,2.1,-0.45,-1.1,-0.53,-0.48,0.13,-0.07000000000000001,-0.36,-2.05,-0.75,-1.58,-0.29,-0.6899999999999999,-0.17,-0.22,0.92,-1.11,-0.66,-0.87,-0.13,0.31],[5.16,3.88,3.51,3.11,2.71,0.77,2.13,0.55,1.53,1.67,2.52,2.63,2.48,3.4,2.69,2.32,4.63,2.32,2.69,2.36,0.64,0.02,0.23,-0.25,0.02,-1.24,-1.21,0.27,1.73,-1.06,-0.11,-0.64,0.9399999999999999,1.92,-1.01,0.55,0.03,-0.32,0.29,-0.34,0.12,0.54,1.14,-1.28,0.92,-0.85,-1.42,-2.67,0.84,2.38],[3.01,1.37,3.49,1.98,4.09,-1.14,-0.02,0.3,3.2,0.09,3.57,3.53,2.56,2.33,5.15,3.59,2.62,3.72,4.71,4.08,0.78,0.05,-1.08,-0.01,2.03,-1.11,1.14,-0.61,-2.19,-1.22,-0.05,1.6,0.93,0.83,1.02,-0.48,1.64,-0.63,-1.38,-0.26,1.68,-2.55,0.01,1.96,-0.39,0.91,-1.45,-1.12,-0.97,-0.07000000000000001],[-2.14,-1.26,0.08,0.19,0.19,0.22,-0.65,0.25,-0.47,-1.19,0.12,-1.04,-0.19,-0.97,0.06,-2.11,0.98,0.37,0.72,1.88,-0.4,0.11,-0.59,0.42,-0.55,-1.84,-2.36,-3.12,-2.82,-3.52,-2.93,-3.18,-4.4,-2.77,-3.89,-3.16,-3.24,-3.03,-3.32,-3.42,-1.02,1.1,0.82,0.17,0.04,1.08,0.38,-1.08,0.19,1.58],[-0.25,0.34,-0.12,-0.99,-1.02,0.84,0.85,0.47,-1.27,-1.11,-1.9,0.49,-0.7,1.78,-0.25,-2.11,-0.59,-0.52,0.51,-0.63,0.19,-1.43,0.4,1.16,0.33,-4.01,-2.9,-2.01,-1.9,-2.72,-3.41,-1.71,-3.25,-4.14,-2.73,-2.36,-4.4,-4.51,-3.3,-2.51,-0.9,-1.58,-0.5,-0.62,-0.68,-0.22,0.39,-0.8100000000000001,-0.55,1.21],[-0.42,0.16,-1.57,1.47,-1.5,-1.15,-2.02,-0.68,0.51,0.75,0.18,-0.44,0.9399999999999999,0.89,-1.69,-0.71,-0.33,0.98,-0.99,0.27,-0.73,1.38,-0.6,-0.27,0.41,-2.15,-2.48,-1.42,-3.11,-3.48,-5.38,-3.64,-2.96,-3.11,-4.87,-3.08,-3.32,-2.88,-2.47,-3.2,1.66,0.31,0.6,-1.43,-0.47,-0.49,1.03,-1.1,-0.92,1.05],[-1.44,0.53,-1.05,0.09,-0.17,-0.53,0.1,-1.02,0.43,1.5,-0.27,0.87,-1.22,-1.89,-0.82,0.75,0.77,-1.79,0.09,-0.18,-1.47,0.49,-0.4,-0.17,-2.21,-1.99,-2.48,-2.74,-3.39,-3.4,-2.77,-3.61,-2.46,-2.09,-1.59,-3.76,-3.02,-2.86,-3.92,-2.42,0.3,-0.86,-0.8100000000000001,0.47,0.57,1.41,0.06,-0.58,0.27,1.08],[-0.15,-0.64,0.52,1.11,-0.5,0.42,-1.08,-0.63,0.84,-0.06,-0.08,0.42,-1.38,-1.48,0.12,1.2,-0.25,-0.34,-1.81,-1.63,-0.27,-1.79,-0.23,1.78,-0.17,-2.19,-3.97,-4.95,-1.21,-1.69,-1.89,-3.99,-2.47,-3.77,-2.47,-2.86,-4.54,-3.78,-6.05,-2.13,0.19,0.85,-0.5600000000000001,1.2,1.7,-0.06,0.77,-0.03,1.3,-0.71],[-0.77,-1.34,0.41,0.21,-0.22,-1.5,0.24,-0.44,1.09,0.29,-0.59,-1.03,0.72,-0.8100000000000001,0.41,-0.19,-0.41,0.21,0.28,-0.6,-2.42,-2.72,-1.68,-4.45,-4.16,-0.05,0.16,0.9399999999999999,-0.39,-0.5600000000000001,-1.5,-3.48,-3.75,-4.02,-4.05,-3.64,-1.6,-3.07,-3.05,-2.49,0.29,-0.04,2.16,0.25,-0.76,2.03,-0.76,0.54,-2.27,-1.37],[1.05,0.35,1.17,1.62,-0.28,-1.16,0.88,0.24,-0.03,-0.08,-2.05,0.06,-0.49,0.48,0.75,0.17,-0.06,0.64,-1.46,-0.85,-2.49,-2.37,-1.83,-2.61,-3.14,-1.09,-1.67,1.13,0.57,-0.63,-3.92,-2,-3.89,-1.67,-2.66,-1.74,-0.8100000000000001,-1.13,-2.26,-3.79,-0.88,-1.22,1.62,0.5,1.6,0.64,-1.07,-1.71,-1.02,-0.23],[1.24,-0.58,1.03,-0.19,-0.42,-2.39,1.65,-0.77,0.76,1.49,0.66,0.49,-0.25,-0.66,0.5600000000000001,-0.78,0.86,0.93,-0.79,1.36,-2.98,-4.19,-2.66,-2.41,-4.43,1.02,0.32,-0.5,-0.14,-1.43,-2.41,-2.13,-2.54,-4.36,-2.36,-2.93,-2.99,-2.28,-2.91,-3.56,0.01,-0.5600000000000001,1.76,0.17,1.07,1.5,0.66,-0.84,0.39,-0.38],[2.07,-1.03,-0.12,0.17,0.55,-0.76,0.04,-0,-0.22,-1.56,-0.24,0.42,-0.45,-1.4,-0.55,0.25,-1.35,1.29,2.1,-1.08,-2.36,-2.19,-3.37,-3.12,-2.06,-0.22,-1.19,0.6,-0.86,-0.67,-2.24,-2.02,-1.51,-2.19,-1.95,-4.38,-4.17,-2.55,-2.45,-2.83,-0.8100000000000001,0.98,1.17,-0.25,-0.26,1.36,1.05,2.23,-1.47,-1.03],[0.16,-1.24,0.12,1.2,-0.7,0.29,0.44,0.63,-1.16,-0.02,1.04,1.31,-0.08,0.92,-0.24,0.35,-1.34,0.43,-0.8100000000000001,0.6899999999999999,-2.29,-2.72,-4.2,-2.36,-3.61,-1.38,-0.93,0.57,2.6,-0.24,-2.31,-3.15,-2.77,-3.67,-2.97,-2.97,-3.37,-5.21,-2.7,-0.87,1.33,-0.12,-1.47,-0.38,-1.47,1.08,-1.23,-3.06,1.45,0.72]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sample1<\/th>\n      <th>Sample2<\/th>\n      <th>Sample3<\/th>\n      <th>Sample4<\/th>\n      <th>Sample5<\/th>\n      <th>Sample6<\/th>\n      <th>Sample7<\/th>\n      <th>Sample8<\/th>\n      <th>Sample9<\/th>\n      <th>Sample10<\/th>\n      <th>Sample11<\/th>\n      <th>Sample12<\/th>\n      <th>Sample13<\/th>\n      <th>Sample14<\/th>\n      <th>Sample15<\/th>\n      <th>Sample16<\/th>\n      <th>Sample17<\/th>\n      <th>Sample18<\/th>\n      <th>Sample19<\/th>\n      <th>Sample20<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","pageLength":5,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Sample1","targets":1},{"name":"Sample2","targets":2},{"name":"Sample3","targets":3},{"name":"Sample4","targets":4},{"name":"Sample5","targets":5},{"name":"Sample6","targets":6},{"name":"Sample7","targets":7},{"name":"Sample8","targets":8},{"name":"Sample9","targets":9},{"name":"Sample10","targets":10},{"name":"Sample11","targets":11},{"name":"Sample12","targets":12},{"name":"Sample13","targets":13},{"name":"Sample14","targets":14},{"name":"Sample15","targets":15},{"name":"Sample16","targets":16},{"name":"Sample17","targets":17},{"name":"Sample18","targets":18},{"name":"Sample19","targets":19},{"name":"Sample20","targets":20}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>And associated metadata:</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-a91001e7ffd9ca98e1fb" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-a91001e7ffd9ca98e1fb">{"x":{"filter":"top","vertical":false,"filterHTML":"<tr>\n  <td><\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n<\/tr>","caption":"<caption>Metadata Table<\/caption>","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20"],["Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20"],["Case","Case","Case","Case","Case","Case","Case","Case","Case","Case","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control"],["Male","Female","Male","Female","Male","Female","Male","Female","Male","Female","Male","Female","Male","Female","Male","Female","Male","Female","Male","Female"],["DrugA","DrugA","DrugA","DrugA","DrugA","DrugB","DrugB","DrugB","DrugB","DrugB","DrugC","DrugC","DrugC","DrugC","DrugC","DrugD","DrugD","DrugD","DrugD","DrugD"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sample<\/th>\n      <th>Condition<\/th>\n      <th>Gender<\/th>\n      <th>Drug<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","pageLength":5,"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Sample","targets":1},{"name":"Condition","targets":2},{"name":"Gender","targets":3},{"name":"Drug","targets":4}],"order":[],"autoWidth":false,"orderClasses":false,"orderCellsTop":true,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>Let’s reshape our data into long format to make it possible to plot with <code>ggplot2</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Reshape the data into long format for ggplot2. Here we use the tidyverse</span></span>
<span><span class="co"># family of packages but there are many ways to approach this reshaping of data</span></span>
<span><span class="va">expression_long</span> <span class="op">&lt;-</span> <span class="va">expression_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># Convert the matrix to a data frame</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># Move the rownames into a column</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"Gene"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># Here we pivot the table, all columns except for the gene column</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">!</span><span class="va">Gene</span>, names_to <span class="op">=</span> <span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we have a data frame with each row representing the expression of a single gene from a single sample:</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-79e903e9f3c1b9ca7c3c" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-79e903e9f3c1b9ca7c3c">{"x":{"filter":"top","vertical":false,"filterHTML":"<tr>\n  <td><\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none;position: absolute;width: 200px;opacity: 1\">\n      <div data-min=\"-6.05\" data-max=\"5.46\" data-scale=\"2\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n<\/tr>","caption":"<caption>Longform Expression Table<\/caption>","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312","313","314","315","316","317","318","319","320","321","322","323","324","325","326","327","328","329","330","331","332","333","334","335","336","337","338","339","340","341","342","343","344","345","346","347","348","349","350","351","352","353","354","355","356","357","358","359","360","361","362","363","364","365","366","367","368","369","370","371","372","373","374","375","376","377","378","379","380","381","382","383","384","385","386","387","388","389","390","391","392","393","394","395","396","397","398","399","400","401","402","403","404","405","406","407","408","409","410","411","412","413","414","415","416","417","418","419","420","421","422","423","424","425","426","427","428","429","430","431","432","433","434","435","436","437","438","439","440","441","442","443","444","445","446","447","448","449","450","451","452","453","454","455","456","457","458","459","460","461","462","463","464","465","466","467","468","469","470","471","472","473","474","475","476","477","478","479","480","481","482","483","484","485","486","487","488","489","490","491","492","493","494","495","496","497","498","499","500","501","502","503","504","505","506","507","508","509","510","511","512","513","514","515","516","517","518","519","520","521","522","523","524","525","526","527","528","529","530","531","532","533","534","535","536","537","538","539","540","541","542","543","544","545","546","547","548","549","550","551","552","553","554","555","556","557","558","559","560","561","562","563","564","565","566","567","568","569","570","571","572","573","574","575","576","577","578","579","580","581","582","583","584","585","586","587","588","589","590","591","592","593","594","595","596","597","598","599","600","601","602","603","604","605","606","607","608","609","610","611","612","613","614","615","616","617","618","619","620","621","622","623","624","625","626","627","628","629","630","631","632","633","634","635","636","637","638","639","640","641","642","643","644","645","646","647","648","649","650","651","652","653","654","655","656","657","658","659","660","661","662","663","664","665","666","667","668","669","670","671","672","673","674","675","676","677","678","679","680","681","682","683","684","685","686","687","688","689","690","691","692","693","694","695","696","697","698","699","700","701","702","703","704","705","706","707","708","709","710","711","712","713","714","715","716","717","718","719","720","721","722","723","724","725","726","727","728","729","730","731","732","733","734","735","736","737","738","739","740","741","742","743","744","745","746","747","748","749","750","751","752","753","754","755","756","757","758","759","760","761","762","763","764","765","766","767","768","769","770","771","772","773","774","775","776","777","778","779","780","781","782","783","784","785","786","787","788","789","790","791","792","793","794","795","796","797","798","799","800","801","802","803","804","805","806","807","808","809","810","811","812","813","814","815","816","817","818","819","820","821","822","823","824","825","826","827","828","829","830","831","832","833","834","835","836","837","838","839","840","841","842","843","844","845","846","847","848","849","850","851","852","853","854","855","856","857","858","859","860","861","862","863","864","865","866","867","868","869","870","871","872","873","874","875","876","877","878","879","880","881","882","883","884","885","886","887","888","889","890","891","892","893","894","895","896","897","898","899","900","901","902","903","904","905","906","907","908","909","910","911","912","913","914","915","916","917","918","919","920","921","922","923","924","925","926","927","928","929","930","931","932","933","934","935","936","937","938","939","940","941","942","943","944","945","946","947","948","949","950","951","952","953","954","955","956","957","958","959","960","961","962","963","964","965","966","967","968","969","970","971","972","973","974","975","976","977","978","979","980","981","982","983","984","985","986","987","988","989","990","991","992","993","994","995","996","997","998","999","1000"],["Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene1","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene2","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene3","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene4","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene5","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene6","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene7","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene8","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene9","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene10","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene11","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene12","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene13","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene14","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene15","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene16","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene17","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene18","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene19","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene20","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene21","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene22","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene23","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene24","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene25","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene26","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene27","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene28","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene29","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene30","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene31","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene32","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene33","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene34","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene35","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene36","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene37","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene38","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene39","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene40","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene41","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene42","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene43","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene44","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene45","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene46","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene47","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene48","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene49","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50","Gene50"],["Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20","Sample1","Sample2","Sample3","Sample4","Sample5","Sample6","Sample7","Sample8","Sample9","Sample10","Sample11","Sample12","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample20"],[-0.96,0.73,0.62,-1.29,-1.33,2.46,1.74,0.6,5.16,3.01,-2.14,-0.25,-0.42,-1.44,-0.15,-0.77,1.05,1.24,2.07,0.16,-0.29,-0.8100000000000001,-0.41,2.64,0.28,4.37,3.76,3.05,3.88,1.37,-1.26,0.34,0.16,0.53,-0.64,-1.34,0.35,-0.58,-1.03,-1.24,0.26,0.27,1.05,0.49,1.09,3.14,2.2,3.34,3.51,3.49,0.08,-0.12,-1.57,-1.05,0.52,0.41,1.17,1.03,-0.12,0.12,-1.15,-1.74,0.6,0.85,-1.64,2.22,2.62,3.43,3.11,1.98,0.19,-0.99,1.47,0.09,1.11,0.21,1.62,-0.19,0.17,1.2,0.2,-1.41,1.02,1.09,-0.46,4.88,1.63,3.21,2.71,4.09,0.19,-1.02,-1.5,-0.17,-0.5,-0.22,-0.28,-0.42,0.55,-0.7,3.03,2.55,3.61,3.23,4.46,0.55,3.52,0.11,0.77,-1.14,0.22,0.84,-1.15,-0.53,0.42,-1.5,-1.16,-2.39,-0.76,0.29,3.09,1.96,3.21,3.07,1.33,-1.84,-1.21,-1.99,2.13,-0.02,-0.65,0.85,-2.02,0.1,-1.08,0.24,0.88,1.65,0.04,0.44,4.12,4.36,1.1,2.02,4.56,-0.55,0.19,1.51,0.55,0.3,0.25,0.47,-0.68,-1.02,-0.63,-0.44,0.24,-0.77,-0,0.63,1.78,3.92,2.32,1.69,1.51,-0.03,0.09,0.35,1.53,3.2,-0.47,-1.27,0.51,0.43,0.84,1.09,-0.03,0.76,-0.22,-1.16,4.27,2.21,3.48,5.46,1.82,-2.16,-0.88,0.43,1.67,0.09,-1.19,-1.11,0.75,1.5,-0.06,0.29,-0.08,1.49,-1.56,-0.02,2.26,3.57,2.54,2.33,2.64,4.77,3.26,3.77,2.52,3.57,0.12,-1.9,0.18,-0.27,-0.08,-0.59,-2.05,0.66,-0.24,1.04,1.87,3.92,2.72,3.91,2.08,3.76,3.66,3.41,2.63,3.53,-1.04,0.49,-0.44,0.87,0.42,-1.03,0.06,0.49,0.42,1.31,2.28,3.26,2.59,3.96,3.85,2.52,3.77,1.02,2.48,2.56,-0.19,-0.7,0.9399999999999999,-1.22,-1.38,0.72,-0.49,-0.25,-0.45,-0.08,3.25,3.35,4.62,4.61,2.51,2.53,0.76,1.91,3.4,2.33,-0.97,1.78,0.89,-1.89,-1.48,-0.8100000000000001,0.48,-0.66,-1.4,0.92,3.15,4.17,2.28,4.84,3.73,3.33,2.1,4.06,2.69,5.15,0.06,-0.25,-1.69,-0.82,0.12,0.41,0.75,0.5600000000000001,-0.55,-0.24,2.69,2.52,2.55,3.7,2.38,2.03,2.83,5.12,2.32,3.59,-2.11,-2.11,-0.71,0.75,1.2,-0.19,0.17,-0.78,0.25,0.35,2.05,2.58,3.01,4.22,3.34,2.47,2.93,2.49,4.63,2.62,0.98,-0.59,-0.33,0.77,-0.25,-0.41,-0.06,0.86,-1.35,-1.34,2.35,3.96,3.22,1.88,4.81,3.06,2.49,3.41,2.32,3.72,0.37,-0.52,0.98,-1.79,-0.34,0.21,0.64,0.93,1.29,0.43,4.22,1.71,3.19,3.67,4.45,3.1,3.52,5.04,2.69,4.71,0.72,0.51,-0.99,0.09,-1.81,0.28,-1.46,-0.79,2.1,-0.8100000000000001,3.2,3.19,2.95,4.22,3.41,4.17,3.04,2.81,2.36,4.08,1.88,-0.63,0.27,-0.18,-1.63,-0.6,-0.85,1.36,-1.08,0.6899999999999999,-0.58,-0.03,-1.5,0.23,0.2,0.39,-0.91,1.37,0.64,0.78,-0.4,0.19,-0.73,-1.47,-0.27,-2.42,-2.49,-2.98,-2.36,-2.29,-0.9399999999999999,0.47,0.37,-0.42,1.68,-0.13,0.91,-0.52,0.02,0.05,0.11,-1.43,1.38,0.49,-1.79,-2.72,-2.37,-4.19,-2.19,-2.72,-0.2,1.02,0.52,0.24,-0.96,-0.38,-0.34,-0.72,0.23,-1.08,-0.59,0.4,-0.6,-0.4,-0.23,-1.68,-1.83,-2.66,-3.37,-4.2,-1.67,0.27,-0.48,-0.55,0.66,1.28,0.09,-0.18,-0.25,-0.01,0.42,1.16,-0.27,-0.17,1.78,-4.45,-2.61,-2.41,-3.12,-2.36,-0.48,0.23,0.67,-0.5,-0.55,-0.7,2.13,-0.4,0.02,2.03,-0.55,0.33,0.41,-2.21,-0.17,-4.16,-3.14,-4.43,-2.06,-3.61,-0.74,0.75,-0.76,1.16,2.44,0.5600000000000001,0.2,0.83,-1.24,-1.11,-1.84,-4.01,-2.15,-1.99,-2.19,-0.05,-1.09,1.02,-0.22,-1.38,1.16,1.22,0.39,2.16,-0.3,0.3,-0.97,-0.79,-1.21,1.14,-2.36,-2.9,-2.48,-2.48,-3.97,0.16,-1.67,0.32,-1.19,-0.93,1.01,0.38,-0.66,-1.71,-0.12,0.22,-0.28,-0.9399999999999999,0.27,-0.61,-3.12,-2.01,-1.42,-2.74,-4.95,0.9399999999999999,1.13,-0.5,0.6,0.57,-0.07000000000000001,-0.99,-1.72,-1.6,0.97,0.92,0.16,0.42,1.73,-2.19,-2.82,-1.9,-3.11,-3.39,-1.21,-0.39,0.57,-0.14,-0.86,2.6,-1.14,-0.16,1.16,-1.04,0.59,-0.89,-1.11,2.1,-1.06,-1.22,-3.52,-2.72,-3.48,-3.4,-1.69,-0.5600000000000001,-0.63,-1.43,-0.67,-0.24,0.9,1.74,0.6899999999999999,0.32,-1.21,-0,-1.93,-0.45,-0.11,-0.05,-2.93,-3.41,-5.38,-2.77,-1.89,-1.5,-3.92,-2.41,-2.24,-2.31,0.85,-0.35,0.14,-0.89,-0.3,-0.2,-0.47,-1.1,-0.64,1.6,-3.18,-1.71,-3.64,-3.61,-3.99,-3.48,-2,-2.13,-2.02,-3.15,0.73,0.6899999999999999,1.49,0.39,0.4,-2.23,0.53,-0.53,0.9399999999999999,0.93,-4.4,-3.25,-2.96,-2.46,-2.47,-3.75,-3.89,-2.54,-1.51,-2.77,0.74,1.22,-1.63,0.24,-1.69,0.48,-0.92,-0.48,1.92,0.83,-2.77,-4.14,-3.11,-2.09,-3.77,-4.02,-1.67,-4.36,-2.19,-3.67,-0.35,0.79,0.13,-0.43,-0.71,1.62,0.68,0.13,-1.01,1.02,-3.89,-2.73,-4.87,-1.59,-2.47,-4.05,-2.66,-2.36,-1.95,-2.97,0.71,-0.01,-2.4,-0.55,-0.25,-1.2,1.52,-0.07000000000000001,0.55,-0.48,-3.16,-2.36,-3.08,-3.76,-2.86,-3.64,-1.74,-2.93,-4.38,-2.97,1.3,0.22,1.44,-1.32,-1,1.66,0.19,-0.36,0.03,1.64,-3.24,-4.4,-3.32,-3.02,-4.54,-1.6,-0.8100000000000001,-2.99,-4.17,-3.37,0.04,-0.89,-0.88,0.68,-0.92,0.2,1.7,-2.05,-0.32,-0.63,-3.03,-4.51,-2.88,-2.86,-3.78,-3.07,-1.13,-2.28,-2.55,-5.21,-0.98,0.44,-1.31,2.16,0.44,1.39,0.28,-0.75,0.29,-1.38,-3.32,-3.3,-2.47,-3.92,-6.05,-3.05,-2.26,-2.91,-2.45,-2.7,0.79,-0.89,-0.88,-0.42,-0.33,-0.86,-0.57,-1.58,-0.34,-0.26,-3.42,-2.51,-3.2,-2.42,-2.13,-2.49,-3.79,-3.56,-2.83,-0.87,0.79,-0.85,-1.16,-1.36,-0.27,0.6899999999999999,-0.16,-0.29,0.12,1.68,-1.02,-0.9,1.66,0.3,0.19,0.29,-0.88,0.01,-0.8100000000000001,1.33,-0.31,-0.99,-1.98,-0.67,1.28,-0.58,2.49,-0.6899999999999999,0.54,-2.55,1.1,-1.58,0.31,-0.86,0.85,-0.04,-1.22,-0.5600000000000001,0.98,-0.12,1.7,-0.65,-0.99,0.65,0.3,-0.68,1.07,-0.17,1.14,0.01,0.82,-0.5,0.6,-0.8100000000000001,-0.5600000000000001,2.16,1.62,1.76,1.17,-1.47,-0.79,1.05,-0.15,0.77,1.27,0.38,0.34,-0.22,-1.28,1.96,0.17,-0.62,-1.43,0.47,1.2,0.25,0.5,0.17,-0.25,-0.38,0.35,-0.39,0.91,2.68,1.03,1.31,0.22,0.92,0.92,-0.39,0.04,-0.68,-0.47,0.57,1.7,-0.76,1.6,1.07,-0.26,-1.47,-2.27,-0.07000000000000001,0.41,-1.37,0.67,0.31,0.07000000000000001,-1.11,-0.85,0.91,1.08,-0.22,-0.49,1.41,-0.06,2.03,0.64,1.5,1.36,1.08,-0.16,-0.46,-1.24,0.06,-0.31,-0.5,0.31,-0.66,-1.42,-1.45,0.38,0.39,1.03,0.06,0.77,-0.76,-1.07,0.66,1.05,-1.23,1.13,0.54,-0.64,-0.2,0.46,1.25,1.51,-0.87,-2.67,-1.12,-1.08,-0.8100000000000001,-1.1,-0.58,-0.03,0.54,-1.71,-0.84,2.23,-3.06,-0.46,0.93,1.93,-1.26,0.48,-0.3,0.62,-0.13,0.84,-0.97,0.19,-0.55,-0.92,0.27,1.3,-2.27,-1.02,0.39,-1.47,1.45,-0.9,-0.21,0.41,-0.66,2.54,0.8100000000000001,0.09,0.31,2.38,-0.07000000000000001,1.58,1.21,1.05,1.08,-0.71,-1.37,-0.23,-0.38,-1.03,0.72]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Gene<\/th>\n      <th>Sample<\/th>\n      <th>Expression<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"t","pageLength":5,"columnDefs":[{"className":"dt-right","targets":3},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Gene","targets":1},{"name":"Sample","targets":2},{"name":"Expression","targets":3}],"order":[],"autoWidth":false,"orderClasses":false,"orderCellsTop":true,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>We plot using the <code>tile</code> geom from ggplot2.</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">heatmap_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">expression_long</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="va">Gene</span>, fill <span class="op">=</span> <span class="va">Expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"blue"</span>, high <span class="op">=</span> <span class="st">"red"</span>, mid <span class="op">=</span> <span class="st">"white"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_axis.html">guide_axis</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Samples"</span>, y <span class="op">=</span> <span class="st">"Genes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">heatmap_plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<section id="perform-clustering" class="level4"><h4 class="anchored" data-anchor-id="perform-clustering">Perform clustering</h4>
<p>It seems there are some patterns in the data, but at this point the samples and genes are not yet clustered.</p>
<p>To view the structure in the data we will first cluster on samples:</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Perform hierarchical clustering:</span></span>
<span><span class="co">## 1. Compute a sample-wise distance matrix</span></span>
<span><span class="va">dist_matrix_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span>
<span><span class="co">## 2. Perform hierarchical clustering</span></span>
<span><span class="va">hclust_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">dist_matrix_samples</span>, method <span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span></span>
<span><span class="co">## 3. Pull out the order of the samples from the hclust object</span></span>
<span><span class="va">ordered_samples</span> <span class="op">&lt;-</span> <span class="va">hclust_samples</span><span class="op">$</span><span class="va">labels</span><span class="op">[</span><span class="va">hclust_samples</span><span class="op">$</span><span class="va">order</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And then cluster on genes. Note that here we don’t <code><a href="https://rdrr.io/r/base/t.html">t()</a></code> to transpose the matrix, and this cluster on genes instead of samples. These functions expect the variable being clustered to be in the rows of the corresponding matrix.</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Perform hierarchical clustering:</span></span>
<span><span class="co">## 1. Compute a sample-wise distance matrix</span></span>
<span><span class="va">dist_matrix_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span></span>
<span><span class="co">## 2. Perform hierarchical clustering</span></span>
<span><span class="va">hclust_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">dist_matrix_genes</span>, method <span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span></span>
<span><span class="co">## 3. Pull out the order of the genes from the hclust object</span></span>
<span><span class="va">ordered_genes</span> <span class="op">&lt;-</span> <span class="va">hclust_genes</span><span class="op">$</span><span class="va">labels</span><span class="op">[</span><span class="va">hclust_genes</span><span class="op">$</span><span class="va">order</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can reorder the input data as we plot using <code><a href="https://forcats.tidyverse.org/reference/fct_relevel.html">fct_relevel()</a></code></p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot heatmap using ggplot2</span></span>
<span><span class="va">heatmap_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">expression_long</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_relevel.html">fct_relevel</a></span><span class="op">(</span><span class="va">Sample</span>, <span class="va">ordered_samples</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_relevel.html">fct_relevel</a></span><span class="op">(</span><span class="va">Gene</span>, <span class="va">ordered_genes</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">Expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"blue"</span>, high <span class="op">=</span> <span class="st">"red"</span>, mid <span class="op">=</span> <span class="st">"white"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="cn">NULL</span>, x <span class="op">=</span> <span class="st">"Samples"</span>, y <span class="op">=</span> <span class="st">"Genes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_axis.html">guide_axis</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">0</span>, r <span class="op">=</span> <span class="fl">3</span>, b <span class="op">=</span> <span class="fl">3</span>, l <span class="op">=</span> <span class="fl">3</span>, unit <span class="op">=</span> <span class="st">"mm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">heatmap_plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What does <code>plot.margin</code> do?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To improve aesthetics, we carefully control the spacing around plots using the theme element <code>plot.margin</code>. This option is specified by using a function called <code><a href="https://ggplot2.tidyverse.org/reference/element.html">margin()</a></code> and tells ggplot how much space to leave around your plots:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">plot</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">0</span>, r <span class="op">=</span> <span class="fl">3</span>, b <span class="op">=</span> <span class="fl">3</span>, l <span class="op">=</span> <span class="fl">3</span>, unit <span class="op">=</span> <span class="st">"mm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section><section id="metadata-plot" class="level4"><h4 class="anchored" data-anchor-id="metadata-plot">Metadata plot</h4>
<p>Create the metadata plot in ggplot.</p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://eliocamp.github.io/ggnewscale/">ggnewscale</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://nanx.me/ggsci/">ggsci</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Options for legend width and heights</span></span>
<span><span class="va">guide_kwd</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="va">guide_kht</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="co"># Create metadata plot</span></span>
<span><span class="va">metadata_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_relevel.html">fct_relevel</a></span><span class="op">(</span><span class="va">Sample</span>, <span class="va">ordered_samples</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Condition"</span>, fill <span class="op">=</span> <span class="va">Condition</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Here we tell the order of the legends</span></span>
<span>  <span class="fu"><a href="https://nanx.me/ggsci/reference/scale_nejm.html">scale_fill_nejm</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span></span>
<span>    order <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    keywidth <span class="op">=</span> <span class="va">guide_kwd</span>,</span>
<span>    keyheight <span class="op">=</span> <span class="va">guide_kht</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggnewscale</span><span class="fu">::</span><span class="fu"><a href="https://eliocamp.github.io/ggnewscale/reference/new_scale.html">new_scale_fill</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Drug"</span>, fill <span class="op">=</span> <span class="va">Drug</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://nanx.me/ggsci/reference/scale_aaas.html">scale_fill_aaas</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span></span>
<span>    order <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    keywidth <span class="op">=</span> <span class="va">guide_kwd</span>,</span>
<span>    keyheight <span class="op">=</span> <span class="va">guide_kht</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggnewscale</span><span class="fu">::</span><span class="fu"><a href="https://eliocamp.github.io/ggnewscale/reference/new_scale.html">new_scale_fill</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Gender"</span>, fill <span class="op">=</span> <span class="va">Gender</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://nanx.me/ggsci/reference/scale_npg.html">scale_fill_npg</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span></span>
<span>    order <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    keywidth <span class="op">=</span> <span class="va">guide_kwd</span>,</span>
<span>    keyheight <span class="op">=</span> <span class="va">guide_kht</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Limits set the order of the main panel</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_y_discrete</a></span><span class="op">(</span></span>
<span>    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Condition"</span>, <span class="st">"Drug"</span>, <span class="st">"Gender"</span><span class="op">)</span>,</span>
<span>    expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Adjust theme elements to remove redundant text etc</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"right"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_axis.html">guide_axis</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata_plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># We add coord fixed just to fit with legend</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why do we use <code>ggnewscale</code>?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We are using a package called <a href="https://eliocamp.github.io/ggnewscale/"><code>ggnewscale</code></a> to specify three distinct color scales for the seperate <code><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile()</a></code> metadata rows. We then set <code>guide=...</code> within the <code><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile()</a></code> function to specify the order of the legends. If we didn’t do this, the legends would be merged across all metadata categories, which we don’t want.</p>
</div>
</div>
</div>
<p>Merge the metadata and heatmap plots together</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute relative heights of main and metadata plot</span></span>
<span><span class="va">rel_height</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span> <span class="op">/</span> <span class="fl">3</span> <span class="co"># Since we have 3 metadata categories</span></span>
<span></span>
<span><span class="co"># Adjust the formatting to adjust the plot margin, remove the axis text and ticks</span></span>
<span><span class="va">metadata_plot</span> <span class="op">&lt;-</span> <span class="va">metadata_plot</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">0</span>, r <span class="op">=</span> <span class="fl">3</span>, b <span class="op">=</span> <span class="fl">0</span>, l <span class="op">=</span> <span class="fl">3</span>, unit <span class="op">=</span> <span class="st">"mm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co"># Combine plots with patchwork</span></span>
<span><span class="va">combined_plot</span> <span class="op">&lt;-</span> <span class="va">metadata_plot</span> <span class="op">+</span> <span class="va">heatmap_plot</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">rel_height</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Print the combined plot</span></span>
<span><span class="va">combined_plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
</section><section id="add-dendrograms" class="level4"><h4 class="anchored" data-anchor-id="add-dendrograms">Add Dendrograms</h4>
<p>Now we will use <a href="https://andrie.github.io/ggdendro/"><code>ggdendro</code></a> to add dendrograms in the margins.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Another good package for plotting ‘tree’ (aka hierarchical trees) data is <a href="https://bioconductor.org/packages/release/bioc/html/ggtree.html"><code>ggtree</code></a>.</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://andrie.github.io/ggdendro/">ggdendro</a></span><span class="op">)</span></span>
<span><span class="va">dendro_top</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://andrie.github.io/ggdendro/reference/ggdendrogram.html">ggdendrogram</a></span><span class="op">(</span><span class="va">hclust_samples</span>, rotate <span class="op">=</span> <span class="cn">FALSE</span>, theme_dendro <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># This expands the y axis to fill the full panel below, and 5% extra space above</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># We need to null the x to remove the space</span></span>
<span>  <span class="co"># between the dendrogram and the plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co"># Layout</span></span>
<span><span class="va">dendro_top</span> <span class="op">+</span> <span class="va">metadata_plot</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="864"></p>
</figure>
</div>
</div>
</div>
<p>Likewise for genes:</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dendro_side</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://andrie.github.io/ggdendro/reference/ggdendrogram.html">ggdendrogram</a></span><span class="op">(</span><span class="va">hclust_genes</span>, rotate <span class="op">=</span> <span class="cn">TRUE</span>, theme_dendro <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># This expands the x axis to fill the full panel below, and 5% extra space above</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># We need to null the x to remove the space</span></span>
<span>  <span class="co"># between the dendrogram and the plot</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t <span class="op">=</span> <span class="fl">0</span>, r <span class="op">=</span> <span class="fl">3</span>, b <span class="op">=</span> <span class="fl">0</span>, l <span class="op">=</span> <span class="fl">0</span>, unit <span class="op">=</span> <span class="st">"mm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co"># Layout</span></span>
<span><span class="op">(</span><span class="va">heatmap_plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_y_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="va">dendro_side</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.25</span><span class="op">)</span>, guides <span class="op">=</span> <span class="st">"collect"</span>, axes <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Plot alignment tip using <code><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion()</a></code>:
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>What is <code>scale_y_continuous(expand = expansion(mult = c(0, 0.05)))</code> doing?</p>
<p>We use this to adjust the total space the given axis takes up in the panel – either stretching the axis to fill the plot, or leave a slight gap either below/above (for the y-axis) or left/right (for the x-axis). See the function <a href="https://ggplot2.tidyverse.org/reference/expansion.html">documentation</a> for more explanation and examples.</p>
<p>So why do we need this? This is needed because we want to ensure the dendrogram is aligned with the main heatmap plot. In this case it required a bit of tweaking to get the values correct. Feel free to adjust the values or remove these adjustments entirely to see what happens!</p>
</div>
</div>
</div>
</section><section id="put-it-all-together" class="level4"><h4 class="anchored" data-anchor-id="put-it-all-together">Put it all together</h4>
<p>Now we need to combine all of these plots together. We think of each plot as a patch in a grid, and also use the <code><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer()</a></code> function from <code>patchwork</code> to fill in the empty gap in this figure.</p>
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># To help we are going to customize each plot margin now to make sure things fit nicely</span></span>
<span><span class="co"># Note the default for margin is 0 space on all sides</span></span>
<span><span class="va">dendro_top_final</span> <span class="op">&lt;-</span> <span class="va">dendro_top</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.025</span>, <span class="fl">0.025</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata_plot_final</span> <span class="op">&lt;-</span> <span class="va">metadata_plot</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">heatmap_plot_final</span> <span class="op">&lt;-</span> <span class="va">heatmap_plot</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_y_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dendro_side_final</span> <span class="op">&lt;-</span> <span class="va">dendro_side</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The default plot spacer has too much space for our purposes so we set</span></span>
<span><span class="co"># plot.margin to be 0</span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the plot!</span></span>
<span><span class="va">dendro_top_final</span> <span class="op">+</span> <span class="va">ps</span> <span class="op">+</span></span>
<span>  <span class="va">metadata_plot_final</span> <span class="op">+</span> <span class="va">ps</span> <span class="op">+</span></span>
<span>  <span class="va">heatmap_plot_final</span> <span class="op">+</span> <span class="va">dendro_side_final</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.25</span><span class="op">)</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="va">rel_height</span><span class="op">)</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="bonus-challenge" class="level4"><h4 class="anchored" data-anchor-id="bonus-challenge">Bonus Challenge</h4>
<p>Can you add a boxplot summarizing the mean expression of each gene to the right of this plot (in between the dendrogram and the main heatmap)? Think carefully about which data should be input to create this plot. Give it a try before opening up the solution below:</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Bonus Challenge Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>First create the boxplot using the long form expression data. Note that because the boxplot is going to be horizontal, the x-axis represents the numeric expression values, while the y-axis represents the individual genes. Also, don’t forget to reorder the genes</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">genexp_boxplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">expression_long</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Expression</span>, y <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_relevel.html">fct_relevel</a></span><span class="op">(</span><span class="va">Gene</span>, <span class="va">ordered_genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    panel.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Expression"</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">genexp_boxplot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<p>Now we add this into our big patchwork. Note: we also have to add extra spacers since our patchwork has an extra column. Finally, we use the <code><a href="https://patchwork.data-imaginist.com/reference/free.html">free()</a></code> function from patchwork to make sure our expression boxplot label is ‘free’ to move next to the plot. Remove this piece of code to see what happens to the label.</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Also remove the axis text. We kept to verify the gene ordering</span></span>
<span><span class="va">genexp_boxplot_final</span> <span class="op">&lt;-</span> <span class="va">genexp_boxplot</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span></span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># The default plot spacer has too much space for our purposes so we set</span></span>
<span><span class="co"># plot.margin to be 0</span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the plot!</span></span>
<span><span class="va">dendro_top_final</span> <span class="op">+</span> <span class="va">ps</span> <span class="op">+</span> <span class="va">ps</span> <span class="op">+</span></span>
<span>  <span class="va">metadata_plot_final</span> <span class="op">+</span> <span class="va">ps</span> <span class="op">+</span> <span class="va">ps</span> <span class="op">+</span></span>
<span>  <span class="va">heatmap_plot_final</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/free.html">free</a></span><span class="op">(</span><span class="va">genexp_boxplot_final</span>, type <span class="op">=</span> <span class="st">"label"</span>, side <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span> <span class="op">+</span> <span class="va">dendro_side_final</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.25</span>, <span class="fl">0.25</span><span class="op">)</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="va">rel_height</span><span class="op">)</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="approach-2---live-recreation-using-ggplot2" class="level3"><h3 class="anchored" data-anchor-id="approach-2---live-recreation-using-ggplot2">Approach 2 - Live Recreation using <code>ggplot2</code>
</h3>
<p>Recreated live with just the input data and final plot. Note the similarities and differences to the first approach.</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://andrie.github.io/ggdendro/">ggdendro</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">expression_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"Gene"</span><span class="op">)</span></span>
<span><span class="va">expression_long</span> <span class="op">&lt;-</span> <span class="va">expression_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span><span class="op">(</span><span class="va">Sample_ID</span>, <span class="va">Expression</span>, <span class="op">-</span><span class="va">Gene</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dist_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="va">expression_data</span>, method <span class="op">=</span> <span class="st">'euclidean'</span><span class="op">)</span></span>
<span><span class="va">hclust_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">dist_mat</span>, method <span class="op">=</span> <span class="st">'ward.D2'</span><span class="op">)</span></span>
<span><span class="va">dendro_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">hclust_mat</span><span class="op">)</span></span>
<span><span class="co"># Convert the dendrogram object to a data frame for ggplot2</span></span>
<span><span class="va">dendro_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://andrie.github.io/ggdendro/reference/dendro_data.html">dendro_data</a></span><span class="op">(</span><span class="va">dendro_mat</span><span class="op">)</span></span>
<span><span class="va">order_genes</span> <span class="op">&lt;-</span> <span class="va">dendro_mat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/labels.html">labels</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="va">expression_long</span><span class="op">$</span><span class="va">Gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">expression_long</span><span class="op">$</span><span class="va">Gene</span>, levels <span class="op">=</span> <span class="va">order_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the ggplot2 object: Gene dendrogram </span></span>
<span><span class="va">for_gene_dendro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://andrie.github.io/ggdendro/reference/segment.html">segment</a></span><span class="op">(</span><span class="va">dendro_data</span><span class="op">)</span></span>
<span><span class="va">gene_dendrogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">for_gene_dendro</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, xend <span class="op">=</span> <span class="va">xend</span>, yend <span class="op">=</span> <span class="va">yend</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">for_gene_dendro</span><span class="op">$</span><span class="va">x</span><span class="op">)</span><span class="op">+</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">for_gene_dendro</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">*</span><span class="fl">1.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span><span class="co"># Optional: Flip the dendrogram horizontally</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span><span class="co"># Adjust vertical axis</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="va">dist_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">'euclidean'</span><span class="op">)</span></span>
<span><span class="va">hclust_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">dist_mat</span>, method <span class="op">=</span> <span class="st">'ward.D2'</span><span class="op">)</span></span>
<span><span class="va">dendro_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html">as.dendrogram</a></span><span class="op">(</span><span class="va">hclust_mat</span><span class="op">)</span></span>
<span><span class="co"># Convert the dendrogram object to a data frame for ggplot2</span></span>
<span><span class="va">dendro_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://andrie.github.io/ggdendro/reference/dendro_data.html">dendro_data</a></span><span class="op">(</span><span class="va">dendro_mat</span><span class="op">)</span></span>
<span><span class="va">order_samples</span> <span class="op">&lt;-</span> <span class="va">dendro_mat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/labels.html">labels</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the ggplot2 object</span></span>
<span><span class="va">for_sample_dendro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://andrie.github.io/ggdendro/reference/segment.html">segment</a></span><span class="op">(</span><span class="va">dendro_data</span><span class="op">)</span></span>
<span><span class="va">sample_dendrogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">for_sample_dendro</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, xend <span class="op">=</span> <span class="va">xend</span>, yend <span class="op">=</span> <span class="va">yend</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">for_sample_dendro</span><span class="op">$</span><span class="va">x</span><span class="op">)</span><span class="op">+</span><span class="fl">0.5</span><span class="op">)</span>, expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">for_sample_dendro</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">*</span><span class="fl">1.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">expression_long</span><span class="op">$</span><span class="va">Sample_ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">expression_long</span><span class="op">$</span><span class="va">Sample_ID</span>, levels <span class="op">=</span> <span class="va">order_samples</span><span class="op">)</span></span>
<span><span class="va">heatmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">expression_long</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample_ID</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">Expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">'white'</span>, linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">order_genes</span><span class="op">)</span>, labels <span class="op">=</span> <span class="va">order_genes</span>,</span>
<span>                     sec.axis <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/sec_axis.html">sec_axis</a></span><span class="op">(</span><span class="op">~</span> <span class="va">.</span>, breaks <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">order_genes</span><span class="op">)</span>, labels <span class="op">=</span> <span class="va">order_genes</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">'RdBu'</span>, </span>
<span>                       limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">6.5</span>, <span class="fl">6.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.title.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.ticks.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        axis.text.y.right <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">'italic'</span><span class="op">)</span>,</span>
<span>        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># </span></span>
<span><span class="va">metas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Condition'</span>,<span class="st">'Gender'</span>,<span class="st">'Drug'</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">Sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Sample</span>, levels <span class="op">=</span> <span class="va">order_samples</span><span class="op">)</span></span>
<span><span class="va">color_palettes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">'Condition'</span> <span class="op">=</span> <span class="st">'Set1'</span>,</span>
<span>                       <span class="st">'Gender'</span> <span class="op">=</span> <span class="st">'Set2'</span>,</span>
<span>                       <span class="st">'Drug'</span> <span class="op">=</span> <span class="st">'Dark2'</span><span class="op">)</span></span>
<span><span class="va">meta_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">metas</span>, <span class="kw">function</span><span class="op">(</span><span class="va">variable</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">variable</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">'white'</span>, linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="va">color_palettes</span><span class="op">[[</span><span class="va">variable</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, label <span class="op">=</span> <span class="va">variable</span>,</span>
<span>                       sec.axis <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/sec_axis.html">sec_axis</a></span><span class="op">(</span><span class="op">~</span> <span class="va">.</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, label <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.title.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.text.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.title.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          axis.ticks.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, unit <span class="op">=</span> <span class="st">'line'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">meta_plots_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">meta_plots</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">sample_dendrogram</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">meta_plots_full</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">meta_plots_full</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">meta_plots_full</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> </span>
<span>  <span class="va">gene_dendrogram</span> <span class="op">+</span> <span class="va">heatmap</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span><span class="op">)</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">order_genes</span><span class="op">)</span><span class="op">)</span>, guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span>, <span class="st">'line'</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
</section><section id="approach-3---complexheatmap" class="level3"><h3 class="anchored" data-anchor-id="approach-3---complexheatmap">Approach 3 - <code>ComplexHeatmap</code>
</h3>
<p>Another great package for composing complex heatmaps is the aptly named <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/"><code>ComplexHeatmap</code></a>. While some like the customization and syntax of <code>ggplot2</code>, <code>ComplexHeatmap</code> can accomplish this figure with arguably far less effort. See the example below using the same code:</p>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap">ComplexHeatmap</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Note that ComplexHeatmap uses the matrix directly</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  matrix <span class="op">=</span> <span class="va">expression_data</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"Expression"</span>,</span>
<span>  clustering_method_columns <span class="op">=</span> <span class="st">"ward.D2"</span>,</span>
<span>  clustering_method_rows <span class="op">=</span> <span class="st">"ward.D2"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p>Add the column annotation following the instructions provided by the <code>ComplexHeatmap</code> authors <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/heatmap-annotations.html">here</a>.</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Expects sample names in rows</span></span>
<span><span class="va">meta_ht</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="st">"Sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set up the colors</span></span>
<span><span class="va">meta_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"Condition"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Case"</span> <span class="op">=</span> <span class="st">"#BC3C29FF"</span>,</span>
<span>    <span class="st">"Control"</span> <span class="op">=</span> <span class="st">"#0072B5FF"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="st">"Drug"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"DrugA"</span> <span class="op">=</span> <span class="st">"#3B4992FF"</span>,</span>
<span>    <span class="st">"DrugB"</span> <span class="op">=</span> <span class="st">"#EE0000FF"</span>,</span>
<span>    <span class="st">"DrugC"</span> <span class="op">=</span> <span class="st">"#008B45FF"</span>,</span>
<span>    <span class="st">"DrugD"</span> <span class="op">=</span> <span class="st">"#631879FF"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="st">"Sex"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Male"</span> <span class="op">=</span> <span class="st">"#E64B35FF"</span>,</span>
<span>    <span class="st">"Female"</span> <span class="op">=</span> <span class="st">"#4DBBD5FF"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span></span>
<span>  df <span class="op">=</span> <span class="va">meta_ht</span>,</span>
<span>  col <span class="op">=</span> <span class="va">meta_cols</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  matrix <span class="op">=</span> <span class="va">expression_data</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"Expression"</span>,</span>
<span>  clustering_method_columns <span class="op">=</span> <span class="st">"ward.D2"</span>,</span>
<span>  clustering_method_rows <span class="op">=</span> <span class="st">"ward.D2"</span>,</span>
<span>  top_annotation <span class="op">=</span> <span class="va">top_anno</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p>Add the boxplot</p>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">right_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span>Expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/anno_boxplot.html">anno_boxplot</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span>, which <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  matrix <span class="op">=</span> <span class="va">expression_data</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"Expression"</span>,</span>
<span>  clustering_method_columns <span class="op">=</span> <span class="st">"ward.D2"</span>,</span>
<span>  clustering_method_rows <span class="op">=</span> <span class="st">"ward.D2"</span>,</span>
<span>  top_annotation <span class="op">=</span> <span class="va">top_anno</span>,</span>
<span>  right_annotation <span class="op">=</span> <span class="va">right_anno</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="4_qctrouble_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There are many additional parameters and ways to customize <code>ComplexHeatmap</code> plots. Refer to the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html">extensive documentation</a>. In many cases it may be easier to see if you can achieve the type of plot you want with <code>ComplexHeatmap</code> before attempting one with <code>ggplot2</code>. It may help with more rapid data exploration as well since it tends to take far less code to get to a quality figure.</p>


</section></section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/Cancer-Research-Institute\.github\.io\/CRI_bioinformatics_workshop\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../course/3_clustering.html" class="pagination-link" aria-label="Clustering concepts and correlation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Clustering concepts and correlation</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../course/test_newfile.html" class="pagination-link" aria-label="Code blocks for DE">
        <span class="nav-page-text">Code blocks for DE</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>